consensus/service/src/timestamp_validator.rs

```diff
@@ -0,0 +1,274 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+use displaydoc::Display;
+use mc_blockchain_types::Block;
+use mc_common::logger::{log, Logger};
+
+/// Provides logic for validating a timestamp used in consensus
+
+const MAX_TIMESTAMP_AGE: u64 = 30 * 1000; // 30 seconds
+
+/// The maximum allowed skew between the system time and the timestamp. This
+/// allows the system time to be a bit behind the timestamp.
+/// The reason for this is that during consensus multiple nodes will be looking
+/// at the time and those nodes may have skew between their clocks. If the node
+/// with the latest time proposes a value, it's possible that the other nodes
+/// will reject it because the timestamp is in the future.
+///
+/// The 3 seconds is taken from the `signed_at` values from the first 2 million
+/// blocks of the blockchain. Throwing away significant outlier nodes most nodes
+/// were almost always within 3 seconds of each other.
+const ALLOWED_FUTURE_SKEW: u64 = 3 * 1000;
+
+/// Errors related to validating timestamps
+#[derive(Clone, Debug, Display, Eq, PartialEq)]
+pub enum Error {
+    /// The timestamp is too far in the future now: {0}, timestamp: {1}
+    InFuture(u64, u64),
+    /** The timestamp is not newer than the last bock. last_bock timestamp: {0},
+    timestamp: {1} */
+    NotNewerThanLastBlock(u64, u64),
+    /// The timestamp is too far in the past now: {0}, timestamp: {1}
+    TooOld(u64, u64),
+}
+
+pub fn validate_with_logger(
+    timestamp: u64,
+    latest_block: &Block,
+    logger: &Logger,
+) -> Result<(), Error> {
+    validate(timestamp, latest_block).map_err(|e| {
+        log::warn!(logger, "Consensus Value timestamp invalid: {e}");
+        e
+    })
+}
```

Comment:
not sure on the thoughts here, I could keep the logging logic fully in mint tx manager if we don't want a function down this low, it just seemed odd to have mint tx manager to explicitly only log the bad time and nothing else.

Reply:
I think this is fine